### PR TITLE
[ Kernel ] Enable Dynamic Per Token `fp8`

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* turns on dynamic per token quantization for `fp8` 
* enables fallback if cutlass kernels are not supported
* add test cases

RESULTS:

`nm-testing/Meta-Llama-3-8B-Instruct-W8A8-FP8-Channelwise-compressed-tensors`

- GSM with dynamic per token:
```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.753|±  |0.0136|
|     |       |strict-match    |     5|exact_match|↑  |0.756|±  |0.0136|
```

- GSM with dynamic per tensor:
```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.748|±  |0.0137|
|     |       |strict-match    |     5|exact_match|↑  |0.747|±  |0.0138|

```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py [hunk 1]
```diff
@@ -113,7 +113,9 @@ def apply_fp8_linear(
     #   If static, layer.input_scale is scalar and x_scale is input_scale.
 
     if cutlass_fp8_supported:
-        qinput, x_scale = ops.scaled_fp8_quant(input, input_scale)
+        qinput, x_scale = ops.scaled_fp8_quant(input,
+                                               input_scale,
+                                               use_per_token_if_dynamic=True)
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/w8a8_utils.py:118]: Shouldn't `use_per_token_if_dynamic` be set by a scheme or something? I don't see why it should always be true for this case

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py:118]: In what case would we not want to use dynamic per token?

Comment (tlrmchlsmth) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py:118]: @mgoin the hypothesis here is that dynamic per token is an overall win over dynamic per tensor when supported. Higher accuracy, but also easier to fuse RMSNorm + Quant, and fewer dependencies so better parallelization for the quantize kernels overall. Downside is more scales.

We'll need to benchmark, but IMO ok for this PR

File: vllm/_custom_ops.py [hunk 2]
```diff
@@ -328,22 +331,19 @@ def scaled_fp8_quant(
     else:
         output = torch.empty_like(input, dtype=torch.float8_e4m3fn)
     if scale is None:
-        scale = torch.zeros(1, device=input.device, dtype=torch.float32)
-        torch.ops._C.dynamic_scaled_fp8_quant(output, input, scale)
+        if use_per_token_if_dynamic:
+            scale = torch.empty((input.numel() // input.shape[-1], 1),
+                                device=input.device,
+                                dtype=torch.float32)
```
Comment (tlrmchlsmth) [inline vllm/_custom_ops.py:337]: Instead of `(input.numel() // input.shape[-1]` would it be better as `input.shape[0]`?

Comment (robertgshaw2-neuralmagic) [reply vllm/_custom_ops.py:337]: I think the input is 3d [batch, seq_len, hidden]?

Comment (tlrmchlsmth) [reply vllm/_custom_ops.py:337]: Ok, I thought the batch and seq_len would be flattened by this point

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py (hunk 1 above)
Comment (tlrmchlsmth) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py:118]: I wouldn't assume a decent performance hit -- the overheads from the CUTLASS epilogues are quite small (like 3%), and there are advantages to doing per-token when quantizing as well. We'll measure.

Comment (tlrmchlsmth) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py:118]: I changed my mind -- as I am implementing the per-token/per-channel wrapper for `torch._scaled_mm`, I think it would be nicer to grab this from a config somewhere

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py:118]: I tried to pass this from the scheme. However, this became very hard because I needed to check if cutlass is supported in multiple places

I think that deciding this here is the right place

**Review by mgoin — APPROVED** [review]

**Review by simon-mo — APPROVED** [review]

**Comment by robertgshaw2-neuralmagic** [comment]
DO NOT LAND UNTIL VARUNS PR GOES IN FIRST
